scrapy_poet/backend.py

```diff
@@ -0,0 +1,45 @@
+import logging
+
+import attr
+import scrapy
+from scrapy.utils.defer import maybe_deferred_to_future
+from web_poet import HttpRequest
+from web_poet.exceptions import (
+    HttpError,
+    HttpRequestError,
+    RequestBackendError,
+)
+
+from scrapy_poet.utils import scrapy_response_to_http_response
+
+logger = logging.getLogger(__name__)
+
+
+def create_scrapy_backend(backend):
+    async def scrapy_backend(request: HttpRequest):
+        if not isinstance(request, HttpRequest):
+            raise TypeError(
+                f"The request should be 'web_poet.HttpRequest' but received "
+                f"one of type: '{type(request)}'."
+            )
+
+        try:
+            request = scrapy.Request(**attr.asdict(request))
+        except TypeError:
+            raise RequestBackendError(
+                f"The given Request isn't compatible with `scrapy.Request`. "
+                f"Ensure that it doesn't contain extra fields: {request}"
+            )
+
+        deferred = backend(request)
+        deferred_or_future = maybe_deferred_to_future(deferred)
+        try:
+            response = await deferred_or_future
+        except scrapy.exceptions.IgnoreRequest:
+            raise HttpError(f"Additional request ignored: {request}")
+        except Exception:
+            raise HttpRequestError(f"Additional request failed: {request}")
```

Comment:
I am not sure how to best approach unexpected exceptions. Here I am assuming they are requests errors (twisted exceptions, openssl ones, etc.), but they could as well come from downloader middlewares.

By the way, I thought I would need to handle https://github.com/scrapy/scrapy/blob/master/scrapy/spidermiddlewares/httperror.py#L13 here and map it to `HttpResponseError`, but it turns out it is a spider middleware, not a downloader middleware, so it is not executed for additional requests, and the default status code handling by `HttpClient` works.

Reply:
Yeah, this is indeed quite tricky. For example, the retry middleware handles a [bunch of different request errors](https://github.com/scrapy/scrapy/blob/master/scrapy/downloadermiddlewares/retry.py#L131-L134) from twisted but [returns None](https://github.com/scrapy/scrapy/blob/master/scrapy/downloadermiddlewares/retry.py#L124) when max retries are reached.

I wonder if it would be safe to assume that if there was no response (e.g. max retries reached) then we can raise `HttpRequestError`.